packages/hash/src/index.ts

```diff
@@ -2,62 +2,93 @@
 // Inspired by https://github.com/garycourt/murmurhash-js
 // Ported from https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash2.cpp#L37-L86
 
-export default function murmur2(str: string): string {
+const hasTextEncoder = typeof TextEncoder !== 'undefined'
+
+const encoder = hasTextEncoder ? new TextEncoder() : undefined
+
+let bufferLength = -1
+let buffer: ArrayBuffer
+let uint8View: Uint8Array
+let int32View: Int32Array
+
+function encode(input: string): number {
+  // Legacy IE11 support
+  if (hasTextEncoder === false) {
+    const bytes = []
+    for (let i = 0; i < input.length; i++) {
+      const codePoint = input.charCodeAt(i)
+      if (codePoint > 0xff) {
+        bytes.push(codePoint >>> 8)
+        bytes.push(codePoint & 0xff)
+      } else {
+        bytes.push(codePoint)
+      }
+    }
+    uint8View = new Uint8Array(bytes)
+    int32View = new Int32Array(uint8View.buffer, 0, Math.floor(bytes.length / 4))
+
+    return bytes.length
+  }
+
+  if (input.length > bufferLength) {
+    // bufferLength must be a multiple of 4 to satisfy Int32Array constraints
+    bufferLength = input.length + (4 - input.length % 4)
+
+    // buffer.resize() is only available in recent browsers, so we re-allocate
+    // a new buffer and views
+    buffer = new ArrayBuffer(bufferLength)
+    uint8View = new Uint8Array(buffer)
+    int32View = new Int32Array(buffer)
+  }
+
+  return encoder!.encodeInto(input, uint8View).written as number;
+}
+
+export default function murmur2(input: string): string {
```

Comment:
While I was exploring the data a bit more I realized that the hashing function is re-hashing the same strings a lot, and that we could speed it up by caching it in a `Map`.

One note before the results, I was testing with a dataset made of a [bunch of MUI components copy/pasted](https://gist.github.com/romgrk/ade3ee62b4641d5b375127e98d354b5d). For this benchmark I switched to a new dataset that is directly extracted from the [MUI dashboard template](https://next.mui.com/material-ui/getting-started/templates/dashboard/), which I think is more realistic and reflects production setups even better.

Here are the results, which indicate a substantial improvement for that dataset when caching the murmur2 result. The implementation is [very naive](https://github.com/romgrk/js-benchmark/blob/master/test-murmur2.js#L135-L147), and I found that trying to mutate/update the map with more complex rules actually decreased the gains by a lot, probably because adding bookkeeping on each hash iteration is too much work. For comparison I also added a case for the cached version, but with a dataset with absolutely no duplicate, which means the caching code is pure overhead.

![image](https://github.com/user-attachments/assets/6586d540-9de5-4ba5-ab34-ea3669dbf832)

Last note, for this particular dataset, unlike the last dataset, SpiderMonkey shows a small/moderate performance decrease from `original` to `improved`, not sure why. This new dataset seems to have a bit shorter string styles, might explain why. I still think `original` to `improved` is a gain considering that Firefox accounts for less than 3% of the trafic right now.

So to sum this up, caching the hash result is a gain *if* we assume that we're going to see a lot of duplicates following each other, which seems to be the case for MUI cases. I would love to have your opinion on this change.